src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/cluster/WorkflowBalancingService.scala

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2018 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.hyperdrive.trigger.scheduler.cluster
+
+import javax.inject.Inject
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Service
+import za.co.absa.hyperdrive.trigger.models.enums.SchedulerInstanceStatuses
+import za.co.absa.hyperdrive.trigger.models.{SchedulerInstance, Workflow}
+import za.co.absa.hyperdrive.trigger.persistance.WorkflowRepository
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait WorkflowBalancingService {
+
+  def getWorkflowsAssignment(runningWorkflowIds: Iterable[Long], instances: Seq[SchedulerInstance], myInstanceId: Long)
+                            (implicit ec: ExecutionContext): Future[(Seq[Workflow], Boolean)]
+
+  def getMaxWorkflowId()(implicit ec: ExecutionContext): Future[Option[Long]]
+}
+
+@Service
+class WorkflowBalancingServiceImpl @Inject()(workflowRepository: WorkflowRepository) extends WorkflowBalancingService {
+  private val logger = LoggerFactory.getLogger(this.getClass)
+
+  override def getWorkflowsAssignment(runningWorkflowIds: Iterable[Long], instances: Seq[SchedulerInstance], myInstanceId: Long)
+                                     (implicit ec: ExecutionContext): Future[(Seq[Workflow], Boolean)] = {
+    val activeInstances = instances.filter(_.status == SchedulerInstanceStatuses.Active)
+    val myRank = getRank(activeInstances, myInstanceId)
+    logger.info(s"Rebalancing workflows on scheduler instance id = $myInstanceId, rank = $myRank," +
+      s" active instance ids = ${activeInstances.map(_.id).sorted}, retaining workflow ids = ${runningWorkflowIds}")
+    for {
+      (releasedWorkflowsCount, instancesDeletedCount) <- workflowRepository.releaseWorkflowAssignmentsOfDeactivatedInstances()
+      _ = if (releasedWorkflowsCount > 0) {
+        logger.info(s"Scheduler instance id = $myInstanceId released $releasedWorkflowsCount workflows of " +
+          s"$instancesDeletedCount deactivated instances")
+      }
+      allWorkflows <- workflowRepository.getWorkflows()
```

Comment:
Yes in the sense that it fetches all workflows. But it's not the joined workflow, i.e. it only selects name, project, isActive, etc. I'd argue it's not that expensive, but the query could easily be optimized by only selecting the ids (and omitting the sort)

Reply:
Do you think it's worthwhile to only select the two instead of the 6 columns of workflow? Otherwise, I won't do it, because it's not worth the complication.

Reply:
I am not sure if it is really worth but I would do it. 
+  acquiredWorkflows <- workflowRepository.getWorkflowsBySchedulerInstance(myInstanceId) - is it really needed ? I think you can return workflowIdsToAcquire

Reply:
No, because `acquireWorkflowAssignments` does `UPDATE workflow SET scheduler_instance_id = myInstanceId WHERE id IN (workflowIdsToAcquire) AND scheduler_instance_id IS NULL`. So not all `workflowIdsToAcquire` might be updated, therefore `SELECT id FROM workflow WHERE scheduler_instance_id = myInstanceId` is not necessarily the same as `workflowIdsToAcquire`